### PR TITLE
[SPARK-28887][K8S] Executor pod status fix

### PR DESCRIPTION
**Pull request by jinxingwang**
### What changes were proposed in this pull request?
If there is a long running sidecar container running in executor pod. the pod status will still be running even if executor container failed. so driver can't properly kill the executor.

### Why are the changes needed?
Without this change, executor pod may stuck where driver think executor still running base on the pod status, but the executor container already dead. 

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
1. ./dev/run-tests
2. Tested this image in a live k8s cluster.
3. changed already used in internal spark image.

### CODE REVIEW

**Comment by tnachen** [comment]
ok to test

**Comment by jinxingwang** [comment]
ok to test

**Comment by dongjoon-hyun** [comment]
Thank you for making a PR, @jinxingwang . Could you fix the scala style? You can check with `dev/scalastyle`.
```
[error] /home/jenkins/workspace/SparkPullRequestBuilder/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala:25:18: Instead of importing implicits in scala.collection.JavaConversions._, import
[error]     scala.collection.JavaConverters._ and use .asScala / .asJava methods
[error] /home/jenkins/workspace/SparkPullRequestBuilder/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala:25:0: collection.JavaConversions._ should be in group 3rdParty, not spark.
```

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodStates.scala [hunk 1]
```diff
@@ -18,20 +18,20 @@ package org.apache.spark.scheduler.cluster.k8s
 
 import io.fabric8.kubernetes.api.model.Pod
 
-sealed trait ExecutorPodState {
+sealed trait ExecutorState {
```
[the remark this answers is not in the document]
Comment (dongjoon-hyun) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodStates.scala]: Hi, @jinxingwang . Could you make another PR for this renaming?
This PR seems to have two orthogonal themes.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala [hunk 2]
```diff
@@ -42,32 +43,47 @@ object ExecutorPodsSnapshot extends Logging {
...
+      ExecutorPodDeleted(pod)
     } else {
       val phase = pod.getStatus.getPhase.toLowerCase(Locale.ROOT)
       phase match {
         case "pending" =>
-          PodPending(pod)
+          ExecutorPending(pod)
         case "running" =>
-          PodRunning(pod)
+          // Checking executor container status is not terminated
```
Comment (tnachen) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala]: Can we add a test for this?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodStates.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (jinxingwang) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodStates.scala]: thanks. will undo re-naming.

**Comment by jinxingwang** [comment]
ok to test

**Comment by jinxingwang** [comment]
ok to test

**Comment by felixcheung** [comment]
please update the PR description also.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala [hunk 3]
```diff
@@ -59,7 +61,24 @@ object ExecutorPodsSnapshot extends Logging {
         case "pending" =>
           PodPending(pod)
         case "running" =>
-          PodRunning(pod)
+          // TODO: upcoming Kubernenetes feature will make this code redundant
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala]: Hi, @jinxingwang .
- First of all, in the shared repository like Apache Spark, we had better use **IDed** `TODO`. If this is needed, you had better use `TODO(SPARK-XXXXX)` style.
- Second, please use the exact version instead of `upcoming` or `latest` because those words become obsolete very soon especially in K8s domain.
- According to the PR content, it looks like you are saying we can skip this PR technically. Could you be more clear why we need to have this code?

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
I left a few comments on the goal of this PR and TODO usage.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala (hunk 3 above)
Comment (jinxingwang) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala]: We needed because currently there is no solution from k8s yet, and this is a bug.
Also even there is a solution, that will require the users to upgrade k8s.

Comment (dongjoon-hyun) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala]: Then, please enumerate the exact K8s versions in the PR description which you target in this PR. Also, you need to create another JIRA issue to remove this TODO (and your code) later.

Comment (jinxingwang) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala]: Will do, thanks!

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsLifecycleManager.scala [hunk 4]
```diff
@@ -176,6 +177,7 @@ private[spark] class ExecutorPodsLifecycleManager(
 
   private def findExitCode(podState: FinalPodState): Int = {
     podState.pod.getStatus.getContainerStatuses.asScala.find { containerStatus =>
+      containerStatus.getName == DEFAULT_EXECUTOR_CONTAINER_NAME &
```
Comment (holdenk) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsLifecycleManager.scala]: Why this change? It seems like if someone manually registered executors with us this might break. That's not as hypothetical as it sounds, since some folks (like I believe some of the Spark CRDs) launch their own executors.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala [hunk 5]
```diff
@@ -59,7 +61,24 @@ object ExecutorPodsSnapshot extends Logging {
...
+          // https://github.com/kubernetes/enhancements/issues/753
+          // Checking executor container status is not terminated
+          // Pod status can still be running if sidecar container status is running
+          val executorContainerStatusCode = pod.getStatus.getContainerStatuses.asScala.
+            filter(_.getName == DEFAULT_EXECUTOR_CONTAINER_NAME).
```
Comment (holdenk) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala]: Same comment as above. If the container name is generally enforced elsehwere that's ok but if we're depending on Spark being the only one to launch we should see if there is another filter we could do (maybe a tag)?

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorLifecycleTestUtils.scala [hunk 6]
```diff
@@ -30,7 +30,7 @@ object ExecutorLifecycleTestUtils {
       .editOrNewStatus()
         .withPhase("failed")
         .addNewContainerStatus()
-          .withName("spark-executor")
+          .withName("spark-kubernetes-executor")
```
Comment (holdenk) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorLifecycleTestUtils.scala]: Is this name change required based on the filtering that was added above or just for clarity when debugging?

**Review by holdenk — COMMENTED** [review]
Thanks for your first PR to Spark! I've got a few questions to try and understand your changes a bit better :) Glad to have more folks working on the Spark + K8s integration :)

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorLifecycleTestUtils.scala (hunk 6 above)
Comment (jinxingwang) [reply resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorLifecycleTestUtils.scala]: This is needed because of the filtering added above.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala (hunk 5 above)
Comment (jinxingwang) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala]: I think you made a really good point, but as far as I know, there is no identifier of container expect the container name. maybe we have to force people use this default name.

**Comment by tnachen** [comment]
ok to test

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala (hunk 5 above)
Comment (holdenk) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala]: Ok, this is probably a rare case but (for example) the Google K8s Spark folks launch their own containers I believe, so let's try and make not of this in the migration guide.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsLifecycleManager.scala (hunk 4 above)
Comment (holdenk) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsLifecycleManager.scala]: CC @mikaylakonst

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala (hunk 2 above)
Comment (holdenk) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala]: Is the failedExecutorWithRuningSidecar bellow now what you're looking for?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala (hunk 5 above)
Comment (skonto) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala]: @holdenk @jinxingwang one option is to introduce a conf option for the executor container name so it can be tracked in cluster mode in case it is modified by the pod template feature.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala [hunk 7]
```diff
@@ -59,7 +61,24 @@ object ExecutorPodsSnapshot extends Logging {
         case "pending" =>
           PodPending(pod)
         case "running" =>
-          PodRunning(pod)
+          // TODO(SPARK-29023): Kubernenetes 1.17 sidecar container feature will
+          // make this code redundant https://github.com/kubernetes/enhancements/issues/753
+          // Checking executor container status is not terminated
+          // Pod status can still be running if sidecar container status is running
```
Comment (skonto) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala]: What if the side car thing provides a service that we need eg. kerberos renewal, shouldnt we remove the executor pod asap and not wait for the Spark container error to happen? We should emit a log msg at least eg. "Spark executor is running but pod is not healthy". Or we could have a configurable property to allow removal if any container is failing. In addition, Pod restartPolicy is set to `Never` since spark manages errors and since we bypass the K8s deployment pattern we should provide a few more managements capabilities. Thoughts?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala (hunk 5 above)
Comment (holdenk) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala]: That would be a cool, I think we can probably just mention it in the migration guide if we don't, but if we add setting for configuring the container name I'm good with that too.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala (hunk 7 above)
Comment (holdenk) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala]: That's a good point. I don't know how often the sidecar is "critical" to keep running along side Spark so I'll defer to other's judgement here.

Comment (jinxingwang) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala]: The current state is if any container is running, the pod will keep running, so there is no case "Spark executor is running but pod is not healthy" but will happen "pod is healthy but Spark executor container(or any sidecar container) is not running "
My change made "main(executor)" container reflect on the pod status.
Your point is very good, that other sidecar also should be healthy when they should be. My thoughts on it are, sidecars are not as critical as "main" container in some case, and should not necessary cause pod to restart all the time. so I will like it to happen is to let "main(executor)" container determine if a pod should be restarted. which should be main container error out when any of its sidecar container is not healthy. 
Let me know what you guys think. thanks.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsLifecycleManager.scala (hunk 4 above)
Comment (jinxingwang) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsLifecycleManager.scala]: This make sure the exit code shows in the pod status is reflect from the executor container.

**Comment by jinxingwang** [comment]
I am closing this PR, because there maybe a better solution to this problem. 
Such as using the restart policy from K8s. which will restart sidecar containers individually, this may be a better approach.
